### PR TITLE
TINKERPOP-1470: InlineFilterStrategy should try and P.or() has() children in OrSteps.

### PR DESCRIPTION
**Pull request by okram**
https://issues.apache.org/jira/browse/TINKERPOP-1470

Added `InlineFilterStrategy` support for:

```
OrStep[HasStep(x,p1),HasStep(y,p2)]
```

to be rewritten as:

```
HasStep(x,p1.or(p2)))
```

While doing this, I found a bug in `ConnectiveP` around nesting predicates that should be inlined. I added upgrade notes about `ConnectiveP` as well as about hidden step labels (forgot to add this from a previous ticket).

Also, I added this ticket as future work: https://issues.apache.org/jira/browse/TINKERPOP-1482

VOTE +1

### CODE REVIEW

**Comment by okram** [comment]
I added some other things to this ticket. In particular:

1. `FilterRankStrategy` is now smart about propagating labels "right" to increase the liklihood of good sorts.
2. Fixed a bug in `TraversalHelper.copyLabels()`.
3. `InlineFilterStrategy` now compresses `HasStep`-chains.

**Comment by okram** [comment]
Ran integration tests overnight.

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 05:12 h
[INFO] Finished at: 2016-09-29T21:09:28-06:00
[INFO] Final Memory: 121M/1519M
[INFO] ------------------------------------------------------------------------
```

**Comment by twilmes** [comment]
I like how you pulled the chunks of processing out into separate methods in `InlineFilterStrategy`.  Cleans things up nicely.

VOTE: +1

**Comment by okram** [comment]
Could you review this @dkuppitz? Check out `FilterRankStrategy(Test)` and tell me what you think.

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/InlineFilterStrategy.java [hunk 1]
```diff
@@ -72,97 +75,163 @@ public void apply(final Traversal.Admin<?, ?> traversal) {
...
-            // filter(x.y) --> x.y
-            for (final TraversalFilterStep<?> step : TraversalHelper.getStepsOfClass(TraversalFilterStep.class, traversal)) {
-                final Traversal.Admin<?, ?> childTraversal = step.getLocalChildren().get(0);
-                if (TraversalHelper.hasAllStepsOfClass(childTraversal, FilterStep.class) &&
-                        !TraversalHelper.hasStepOfClass(childTraversal,
-                                DropStep.class,
-                                RangeGlobalStep.class,
-                                DedupGlobalStep.class,
-                                LambdaHolder.class)) {
+            for (final FilterStep<?> step : TraversalHelper.getStepsOfAssignableClass(FilterStep.class, traversal)) {
```
Comment (dkuppitz) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/InlineFilterStrategy.java]: Complex traversal could waste a few CPU cycles in this method. Here's an optimized version:

```
    @Override
    public void apply(final Traversal.Admin<?, ?> traversal) {
        boolean changed = true; // recursively walk child traversals trying to inline them into the current traversal line.
        while (changed) {
            changed = false;
            final Iterator<FilterStep> filterStepIterator = TraversalHelper.getStepsOfAssignableClass(FilterStep.class, traversal).iterator();
            while (!changed && filterStepIterator.hasNext()) {
                final FilterStep<?> step = filterStepIterator.next();
                changed = step instanceof HasStep && InlineFilterStrategy.processHasStep((HasStep) step, traversal) ||
                        step instanceof TraversalFilterStep && InlineFilterStrategy.processTraversalFilterStep((TraversalFilterStep) step, traversal) ||
                        step instanceof OrStep && InlineFilterStrategy.processOrStep((OrStep) step, traversal) ||
                        step instanceof AndStep && InlineFilterStrategy.processAndStep((AndStep) step, traversal);
            }
            if (!changed && traversal.getParent() instanceof EmptyStep) {
                final Iterator<MatchStep> matchStepIterator = TraversalHelper.getStepsOfClass(MatchStep.class, traversal).iterator();
                while (!changed && matchStepIterator.hasNext()) {
                    if (InlineFilterStrategy.processMatchStep(matchStepIterator.next(), traversal))
                        changed = true;
                }
            }
        }
    }
```

**Comment by dkuppitz** [comment]
LGTM.

VOTE: +1

**Comment by okram** [comment]
merged.

**Pull request closed without merging**